include/yamail/resource_pool/async/detail/pool_impl.hpp

```diff
@@ -60,6 +60,47 @@ auto make_on_list_iterator_handler(const boost::system::error_code& error, ListI
     return result_type(error, list_iterator, std::forward<Handler>(handler));
 }
 
+template <class ListIterator>
+struct base_list_iterator_handler_impl {
+    virtual void operator ()(boost::system::error_code ec, ListIterator iterator) = 0;
+    virtual ~base_list_iterator_handler_impl() = default;
+};
+
+template <class ListIterator, class Handler>
+class list_iterator_handler_impl final : public base_list_iterator_handler_impl<ListIterator> {
+public:
+    template <class HandlerT>
+    list_iterator_handler_impl(HandlerT&& handler)
+            : handler(std::forward<HandlerT>(handler)) {
+        static_assert(std::is_same<std::decay_t<HandlerT>, Handler>::value, "HandlerT is not Handler");
+    }
+
+    void operator ()(boost::system::error_code ec, ListIterator iterator) final {
+        handler(ec, iterator);
+    }
+
+private:
+    Handler handler;
+};
+
+template <class ListIterator>
+class list_iterator_handler {
+public:
+    list_iterator_handler() = default;
+
+    template <class Handler>
+    list_iterator_handler(Handler&& handler)
+        : impl(std::make_unique<list_iterator_handler_impl<ListIterator, Handler>>(std::forward<Handler>(handler))) {
```

Comment:
I made a utility function to eliminate a boilerplate code of template specialization for [std::allocate_shared](https://github.com/yandex/ozo/blob/master/include/ozo/failover/strategy.h#L367-L371) so you may do the same thing to simplify such case.

Reply:
It's only a single place now. I don't consider this as a boilerplate.